### PR TITLE
Import projects

### PR DESCRIPTION
**Pull request by alanvardy**

### CODE REVIEW

**Comment by alanvardy** [comment]
Current dependencies on/for this PR:
* main
  * **PR #356** <a href="https://app.graphite.dev/github/pr/alanvardy/tod/356" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/alanvardy/tod/356?utm_source=stack-comment).

File: src/config.rs [hunk 1]
```diff
@@ -212,30 +203,19 @@ pub fn get_or_create(config_path: Option<String>) -> Result<Config, String> {
...
-    let home_directory = dirs::home_dir()
-        .ok_or_else(|| String::from("Could not find home directory"))?
-        .to_str()
-        .ok_or_else(|| String::from("Could not convert directory to string"))?
-        .to_owned();
-    Ok(format!("{home_directory}/{filename}"))
+    if cfg!(test) {
+        _ = fs::create_dir(format!("{config_directory}/tod_test"));
+        let random_string = Alphanumeric.sample_string(&mut rand::thread_rng(), 30);
+        Ok(format!("tests/{random_string}.testcfg"))
```
Comment (alanvardy) [inline src/config.rs:215]: So now EVERY config created in test gets it's own unique file so testing should be a lot easier. Just create a config and keep referencing that same path for the rest of the test.

It is created in the test directory and gitignored

File: src/projects.rs [hunk 2]
```diff
@@ -67,6 +98,33 @@ fn fetch_next_item(config: &Config, project_name: &str) -> Result<Option<Item>,
     Ok(items.first().map(|item| item.to_owned()))
 }
 
+/// Fetch projects and prompt to add them to config one by one
+pub fn import(config: Config) -> Result<String, String> {
```
Comment (titoOdUA) [inline src/projects.rs:102]: This function is doing a lot(preparing prompts, filtering projects, iterating over projects, collecting the input, and modifying the config). Maybe we can split it into some logical parts.

File: src/projects.rs [hunk 3]
```diff
@@ -67,6 +98,33 @@ fn fetch_next_item(config: &Config, project_name: &str) -> Result<Option<Item>,
...
+    let new_projects: Vec<Project> = projects
+        .into_iter()
+        .filter(|p| !project_ids.contains(&p.id))
+        .collect();
+    for project in new_projects {
+        println!("{}", project.fmt());
+        match config::select_input("Select an option", options.clone()) {
+            Ok(string) => {
+                if string == "add" {
+                    let modified_config = Config::load(&config.path)?;
```
Comment (titoOdUA) [inline src/projects.rs:115]: why do we need load config if we provide one as a parameter?
maybe just provide the mutable one?

**Review by titoOdUA — COMMENTED** [review]
Hi, I left some comments. Maybe you would find one of them useful.
Other than that it looks solid.

File: src/projects.rs (hunk 3 above)
Comment (alanvardy) [reply src/projects.rs:115]: Haha, this is a great callout. I'm used to working with Elixir where there is no mutability so up until now I have just been using return values, but this is an excellent example of where that strategy falls down.

Fixed to make config mutable, compiler is a huge help.

File: src/projects.rs (hunk 2 above)
Comment (alanvardy) [reply src/projects.rs:103]: Absolutely, split out a few things

File: src/projects.rs [hunk 4]
```diff
@@ -2,9 +2,40 @@ use crate::config::Config;
...
+    pub is_shared: bool,
+    pub is_favorite: bool,
+    pub is_inbox_project: bool,
+    pub is_team_inbox: bool,
+    pub view_style: String,
+    pub url: String,
+    pub parent_id: Option<String>,
+}
+
+impl Project {
```
[the remark this answers is not in the document]
Comment (alanvardy) [reply src/projects.rs:26]: Yeah, great idea :+1: I'm unfamiliar with traits so thanks for calling this out

**Comment by alanvardy** [comment]
I appreciate these reviews @titoOdUA , you're giving me lots to improve upon!

File: src/projects.rs [hunk 5]
```diff
@@ -67,6 +101,46 @@ fn fetch_next_item(config: &Config, project_name: &str) -> Result<Option<Item>,
     Ok(items.first().map(|item| item.to_owned()))
 }
 
+/// Fetch projects and prompt to add them to config one by one
+pub fn import(config: Config) -> Result<String, String> {
+    let projects = request::projects(&config)?;
+    let new_projects = filter_new_projects(&config, projects);
+    let mut config = config;
```
Comment (titoOdUA) [inline src/projects.rs:108]: we probably could get reed of this sort of reassignment by changing the signatures of some functions

but it is so minor that I don't know if it is beneficial to do at all. just bugs me a little =)

```rust
#[cfg(not(tarpaulin_include))]
fn project_import(matches: &ArgMatches) -> Result<String, String> {
    let mut config = fetch_config(matches)?;

    projects::import(&mut config)
}

//change import fn to receive mut ref
pub fn import(config: &mut Config) -> Result<String, String> {
    let projects = request::projects(config)?;
    let new_projects = filter_new_projects(config, projects);
    for project in new_projects {
        maybe_add_project(config, project)?;
    }
    Ok(green_string("No more projects"))
}
```

and change a few existing assignments 
note that way in import fn
```rust
    let projects = request::projects(config)?;
    let new_projects = filter_new_projects(config, projects);
```
we don't strictly need &config, our config is already &mut, and Rust compiler can infer it from the context and signatures of the functions to which this ref is being provided, and "transform our &mut to &"

Just find it interesting. We can opt-in and be explicit but clippy will complain about it by default

**Review by titoOdUA — DISMISSED** [review]

File: src/projects.rs (hunk 5 above)
Comment (alanvardy) [reply src/projects.rs:108]: Cool, thank you. I've been avoiding passing mutable references up until now but it's obviously the better way to do it in Rust. :+1:

**Pull request merged**